File: binary-search-for-the-answer/binary-search-for-the-answer.py
def check_shuffle(mid, ar, k):
    ch = 1
    last_k = ar[0]

    for data in ar:
        if (data - last_k) > mid:
            ch += 1
            last_k = data

    return ch <= k


N = int(input())
k = int(input())

arr_points = []
for i in range(N):
    point = int(input())
    arr_points.append(point)

left = 0
right = arr_points[-1] - arr_points[0]

while left < right:
    middle = (left + right) // 2
    if check_shuffle(middle, arr_points, k):
        right = middle
    else:
        left = middle + 1

print(left)

File: binary-search-for-the-answer/BinarySearchForTheAnswer.java
import java.util.Scanner;
import java.util.stream.IntStream;

public class BinarySearchForTheAnswer {
    private int[] array;

    public static void main(String[] args) {
        BinarySearchForTheAnswer search = new BinarySearchForTheAnswer();
        Scanner scanner = new Scanner(System.in);
        
        int N = Integer.parseInt(scanner.nextLine());
        int K = Integer.parseInt(scanner.nextLine());
        
        search.array = new int[N];
        IntStream.range(0, K).forEach(i -> search.array[i] = Integer.parseInt(scanner.nextLine()));
        System.out.println(search.solve(K));
    }

    private boolean check(int x, int k) {
        
        int lines = 1;
        int last = array[0];
        
        for (int j : array) {
            if (j - last >= x) {
                lines++;
                last = j;
            }
        }
        return lines > k;
    }

    public int solve(int k) {
        int left = 0;
        int right = (array[array.length-1] - array[0]) + 1;
        while (right - left > 1) {
            int mid = (left + right) / 2;
            if(check(mid, k)) {
                left = mid;
            } else {
                right = mid;
            }
        }
        return left;
    }

}
